### PR TITLE
[Coroutines] Make CoroElide actually work for (trivial) C++ coroutines

### PR DESCRIPTION
**Pull request by yuxuanchen1997**
## Motivation
`Task` is widely used in C++ coroutines. We have various reports that `CoroElide` are not working in coroutines that actually `co_await` a `Task`. Further inspection revealed there are multiple issues. 

Consider this most trivial task type. 
```c++
struct Task {
  struct promise_type {
    struct FinalAwaiter {
      bool await_ready() const noexcept { return false; }

      template <typename P>
      std::coroutine_handle<> await_suspend(std::coroutine_handle<P> coro) noexcept {
        if (!coro)
          return std::noop_coroutine();
        return coro.promise().continuation;
      }
      void await_resume() noexcept {}
    };

    Task get_return_object() noexcept {
      return std::coroutine_handle<promise_type>::from_promise(*this);
    }

    std::suspend_always initial_suspend() noexcept { return {}; }
    FinalAwaiter final_suspend() noexcept { return {}; }
    void unhandled_exception() noexcept {}
    void return_value(int x) noexcept {
      value = x;
    }

    std::coroutine_handle<> continuation;
    int value;
  };

  Task(std::coroutine_handle<promise_type> handle) : handle(handle) {}
  ~Task() {
    if (handle)
      handle.destroy();
  }

  struct Awaiter {
    Awaiter(Task *t) : task(t) {}
    bool await_ready() const noexcept { return false; }
    void await_suspend(std::coroutine_handle<void> continuation) noexcept {}
    int await_resume() noexcept {
      return 43;
    }

    Task *task;
  };

  auto operator co_await() {
    return Awaiter{this};
  }

private:
  std::coroutine_handle<promise_type> handle;
};
```

This patch addresses two most obvious issues in `CoroElide` that prevents Elide from happening in simple inlined `co_await` cases. 

1. CoroElide analysis requires that a coroutine frame pointer gets destroyed through `subfn.addr` (which is generally just `handle.destroy()`. However, in C++, the pointer is often first stored to `Task` and then retrieved from the `Task` struct before destruction. This maps into the code as in the constructor and destructor of the `Task` type. This patch RAUW all such uses.
2. Make `canCoroBeginEscape` stronger with more analysis for its predecessor block.

TODO:
1. Task type's storage for coroutine handle needs to be constant for every instance of `Task`. Currently this is not true for move contructors. Make the pass only RAUW if the task has not been tampered with (e.g. with move, awaiter::await_suspend)
3. Revisit `getNullCheckerPredecessor`. It works well enough for the following pattern 
```
if (handle)
  handle.destroy()
```
However, we need to think about more robust ways other than brute force pattern matching. 
4. Add unit tests.

### CODE REVIEW

File: llvm/lib/Transforms/Coroutines/CoroElide.cpp [hunk 1]
```diff
@@ -193,23 +195,51 @@ CoroIdElider::CoroIdElider(CoroIdInst *CoroId, FunctionElideInfo &FEI,
       CoroAllocs.push_back(CA);
   }
 
+  DenseMap<AllocaInst *, CoroBeginInst *> Writes;
+  for (CoroBeginInst *CB : CoroBegins) {
+    for (User *U : CB->users()) {
+      if (auto *Store = dyn_cast<StoreInst>(U)) {
```
Comment (apolloww) [inline llvm/lib/Transforms/Coroutines/CoroElide.cpp:201]: Have we seen case of the alloca stores into another alloca?

I think we can use a worklist here to catch all the allocas that alias with the corobegin.

Comment (yuxuanchen1997) [reply llvm/lib/Transforms/Coroutines/CoroElide.cpp:201]: This alloca is clang generated. (usually just the temporary value of type `Task`.) I am thinking that if we can somehow annotate from the front end which field in a `Task` is a handle, we can potentially avoid having to do a convoluted analysis that's not easy to get right. 

Are you asking whether this `%t = alloca %Task` gets `store`d into another `alloca`ed stack memory? Yes it does. Usually operator co_await does that to create an `Awaiter` object.

File: llvm/lib/Transforms/Coroutines/CoroElide.cpp [hunk 2]
```diff
@@ -253,6 +283,44 @@ void CoroIdElider::elideHeapAllocations(uint64_t FrameSize, Align FrameAlign) {
   removeTailCallAttribute(Frame, AA);
 }
 
+static BasicBlock *getNullCheckerPredecessor(Value *Handle,
```
Comment (WenleiHe) [inline llvm/lib/Transforms/Coroutines/CoroElide.cpp:286]: > However, we need to think about more robust ways other than brute force pattern matching.

I actually that the way it's handle is okay. At least I don't see correctness issue IIUC. 

To formulate the problem/solution, we are essentially trying to mark paths safe/non-escape that are either 1) destroyed, or 2) null. null is okay obviously as nothing is escaping. The way you implement this is by hoisting the the safe/visited marking to the dominator of 1) and 2), not the most elegant, but does its job IMHO.

**Comment by WenleiHe** [comment]
Has the work list search limit (`Limit = 32 * (1 + It->second.size())`) for escape ever turn out to be a problem?

**Pull request closed without merging**